src/sage/rings/polynomial/infinite_polynomial_ring.py

```diff
@@ -922,7 +931,7 @@ def _element_constructor_(self, x):
             if isinstance(self._base, MPolynomialRing_polydict):
                 x = sage_eval(repr(), next(self.gens_dict()))
             else:
-                x = self._base(x)
+                x = self._base.coerce(x)
```

Comment:
I find this change to be slightly dangerous. The element construction is the conversion, but now you are requiring the input to satisfy a coercion into the base. So it might not be able to convert rational numbers into integers (e.g., `2/1`). Please add some appropriate examples showing that `InfinitePolynomialRing(ZZ, 'x')` works as expected with integers realized as rational numbers and similar polynomials.

Reply:
Do you meant as follows:
```
             sage: R = InfinitePolynomialRing(ZZ, "a")
             sage: R(GF(5)(2))
             2
```
?  (There is something odd, because the 2 above has parent `ZZ`,  not `InfinitePolynomialRing(ZZ, "a")`.

However, the change to `x = self._base.coerce(x)` is actually what resolved the bug
```
            sage: L.<x, y> = QQ[]
            sage: R.<a> = InfinitePolynomialRing(QQ)
            sage: M = InfinitePolynomialRing(L, names=["a"])
            sage: c = a[0]
            sage: M(c)
            a_0
```
Moreover, this is also done in `MPolynomialRing_polydict.__call__`.

Reply:
I agree that a complete rewrite of `_element_constructor_` is needed, but I'm a bit afraid that this is not easy.  In particular, comments like "it's even more of a shame that MPolynomialRing_polydict does not work in complicated settings" don't help improving my confidence.

Reply:
That might be the next thing we try to do after we get the functional solver working. As such, I can accept this if we put a (code) comment saying this should instead do a conversion instead of a coercion as a `# FIXME`.

Also note that I don't think this works if I want to be evil and have my base ring have variable names the same as the inf poly ring.

Reply:
As mentioned in our recent email exchange, we've effectively found an example where this change produces undesirable results.

Reply:
I guess it would be good to make this example available somehow for the afterworld, I am not extremely motivated, though.